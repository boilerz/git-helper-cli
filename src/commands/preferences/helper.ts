import chalk from 'chalk';

// eslint-disable-next-line import/prefer-default-export
export function organisationsSuffix(organisations: string[]) {
  return organisations.length > 0
    ? ` (${chalk.yellow(organisations.join(', '))})`
    : '';
}
