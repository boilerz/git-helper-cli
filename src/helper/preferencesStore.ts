import Conf from 'conf';

export interface Preferences {
  foo: string;
  bar: boolean;
  baz: number;
}

const preferencesStore = new Conf<Preferences>({
  defaults: {
    foo: 'foo',
    bar: true,
    baz: 42,
  },
});

export default preferencesStore;
