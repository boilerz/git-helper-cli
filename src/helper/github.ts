import type { components } from '@octokit/openapi-types';
import { Octokit } from '@octokit/rest';
import chalk from 'chalk';
import ora from 'ora';

import preferencesStore from './preferencesStore';

export type IssueSearchResultItem =
  components['schemas']['issue-search-result-item'];
export type PullRequest = components['schemas']['pull-request'];
const DEFAULT_PER_PAGE = 100;
let octokit: Octokit;

export function setup(): boolean {
  const githubTokenEnvVar = preferencesStore.get('githubTokenEnvVar');
  const token = process.env[githubTokenEnvVar];
  if (!token) {
    console.log(chalk.bold.yellow('🔑 Missing github token.'));
    console.log(
      chalk.italic.blue(
        `Please set the token at ${chalk.yellow(
          githubTokenEnvVar,
        )} or update env var key using ${chalk.red('-p')} option.`,
      ),
    );
    return false;
  }

  octokit = new Octokit({
    auth: token,
  });
  return true;
}

function getUserOrganisationNames(): string[] {
  const organisations = preferencesStore.get('organisations');
  if (organisations.length === 0) {
    console.log(chalk.yellow('No organisations set'));
    return [];
  }
  return organisations;
}

export async function checkOrganisationExistence(
  organisationName: string,
): Promise<boolean> {
  try {
    const { status } = await octokit.orgs.get({ org: organisationName });
    return status === 200;
  } catch (err) {
    return false;
  }
}

async function getPullRequest(
  owner: string,
  repo: string,
  number: number,
): Promise<PullRequest> {
  const { data: pullRequest } = await octokit.pulls.get({
    owner,
    repo,
    pull_number: number,
  });
  return pullRequest;
}

export type MergeableState = 'clean' | 'unstable';

interface MergeOptions {
  mergeableStates?: MergeableState[];
}

export async function mergePullRequest(
  owner: string,
  repo: string,
  number: number,
  { mergeableStates = ['clean'] }: MergeOptions = {},
) {
  const pullRequest = await getPullRequest(owner, repo, number);

  if (
    !pullRequest.mergeable ||
    !mergeableStates.includes(pullRequest.mergeable_state as MergeableState)
  ) {
    console.log(
      chalk.gray(
        `⏩ Will pass for ${pullRequest.html_url}, (state: ${chalk.yellow(
          pullRequest.mergeable_state,
        )})`,
      ),
    );
    return;
  }

  const mergeSpinner = ora(
    `Will try to merge ${owner}/${repo}/${number}`,
  ).start();
  try {
    await octokit.pulls.merge({
      owner,
      repo,
      pull_number: number,
    });
    mergeSpinner.succeed(
      `Merge success (${owner}/${repo}/${number}) - ${pullRequest.mergeable_state}`,
    );
  } catch (err) {
    mergeSpinner.fail(
      `Fail to merge ${owner}/${repo}/${number} => ${err.message}`,
    );
  }
}

export async function getOpenedPullRequests(): Promise<
  IssueSearchResultItem[]
> {
  const organisationNames = getUserOrganisationNames();
  if (!organisationNames.length) return [];

  const organisationFilter = organisationNames
    .map((name) => `user:${name}`)
    .join(' ');
  const {
    data: { items },
  } = await octokit.search.issuesAndPullRequests({
    q: `is:open is:pr archived:false ${organisationFilter}`,
    per_page: DEFAULT_PER_PAGE,
  });
  return items;
}

export async function mergeMergeablePullRequests(
  options?: MergeOptions,
): Promise<void> {
  const pullRequests = await getOpenedPullRequests();
  if (!pullRequests.length) {
    console.log(chalk.yellow('🔀 No PR found'));
    return;
  }

  for (const pullRequest of pullRequests) {
    const [number, , repo, owner] = pullRequest.html_url.split('/').reverse();
    const pullNumber = parseInt(number, 10);
    await mergePullRequest(owner, repo, pullNumber, options);
  }
}
