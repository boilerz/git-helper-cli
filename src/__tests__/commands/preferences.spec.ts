import inquirer from 'inquirer';
import { mocked } from 'ts-jest/utils';

import preferencesCommand from '../../commands/preferences';
import preferencesStore from '../../helper/preferencesStore';
import { displayOutput, runCommand } from '../helper';

const mockedInquirer = mocked(inquirer, true);

describe('commands/preferences', () => {
  it('should display help', async () => {
    const output = await displayOutput(preferencesCommand, '--help');

    expect(output).toMatchInlineSnapshot(`
      "test [command]

      Commands:
        test preferences  Set preferences                                 [aliases: p]

      Options:
        --help     Show help                                                 [boolean]
        --version  Show version number                                       [boolean]"
    `);
  });

  it('should prompt and store preferences', async () => {
    mockedInquirer.prompt.mockResolvedValue({
      foo: 'foo',
      bar: false,
      baz: 8,
    });

    await runCommand(preferencesCommand, 'p');

    expect(inquirer.prompt).toMatchSnapshot();
    expect(preferencesStore.get('foo')).toEqual('foo');
    expect(preferencesStore.get('bar')).toBe(false);
    expect(preferencesStore.get('baz')).toEqual(8);
  });
});
