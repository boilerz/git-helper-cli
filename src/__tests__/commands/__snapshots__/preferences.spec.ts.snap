// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/preferences should prompt and store preferences 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "default": "foo",
          "message": "Choose foo",
          "name": "foo",
          "type": "input",
        },
        Object {
          "default": true,
          "message": "Enable bar",
          "name": "bar",
          "type": "confirm",
        },
        Object {
          "default": 42,
          "message": "Choose baz",
          "name": "baz",
          "type": "number",
        },
      ],
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
